### PR TITLE
Allay Support

### PR DESCRIPTION
**Pull request by Absolutionism**
### Description
This PR aims to add support for Allays. Allowing users to duplicate allays, set if allays can duplicate naturally, make them dance, duplication cooldown, and their target jukebox.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/effects/EffAllayCanDuplicate.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+import org.bukkit.entity.Allay;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Allay Duplicate")
+@Description("Set whether an allay can or cannot duplicate itself.")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffAllayCanDuplicate.java:18]: shuold include that this is different from breeding, just to make sure

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import org.bukkit.entity.Allay;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Make Allay Dance")
+@Description({
+	"Make an allay start or stop dancing.",
+	"By providing a location when making an allay dance, checks to see if the block at the location is a jukebox and playing music.",
+	"By not providing a location, the allay will non-stop dance."
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffAllayDancing.java:24]:
```suggestion
	"By not providing a location, the allay will dance forever."
```

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/conditions/CondIsDancing.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+@Description("Checks to see if an entity is dancing. (e.g allays and parrots)")
+@Examples({
+	"if last spawned allay is dancing:",
+		"\tbroadcast \"Dance Party!\""
+})
+@Since("INSERT VERSION")
+public class CondIsDancing extends PropertyCondition<LivingEntity> {
+
+	static {
+		PropertyCondition.register(CondIsDancing.class, PropertyType.BE, "dancing", "livingentities");
```
Comment (Burbulinis) [inline src/main/java/ch/njol/skript/conditions/CondIsDancing.java:22]:
```suggestion
		PropertyCondition.register(CondIsDancing.class, "dancing", "livingentities");
```
the type is `BE` by default

**Review by Burbulinis — APPROVED** [review]

File: src/main/java/ch/njol/skript/conditions/CondIsDancing.java (hunk 3 above)
Comment (Burbulinis) [reply src/main/java/ch/njol/skript/conditions/CondIsDancing.java:22]: also you can just do `register(...)` instead of `PropertyCondition.register(...)`

Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/conditions/CondIsDancing.java:22]: Hmmm, I'm pretty sure I tried that before at one point and it didn't show. I could be wrong.

File: src/main/java/ch/njol/skript/conditions/CondIsDancing.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import org.bukkit.entity.Allay;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Parrot;
+
+@Name("Is Dancing")
+@Description("Checks to see if an entity is dancing. (e.g allays and parrots)")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/conditions/CondIsDancing.java:13]:
```suggestion
@Description("Checks to see if an entity is dancing, such as allays or parrots.")
```

**Review by Efnilite — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprAllayJukebox.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+	static {
+		registerDefault(ExprAllayJukebox.class, Location.class, "target jukebox", "livingentities");
+	}
+
+	@Override
+	public @Nullable Location convert(LivingEntity entity) {
+		if (!(entity instanceof Allay allay))
+			return null;
+		return allay.getJukebox();
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprAllayJukebox.java:27]:
```suggestion
		return entity instanceof Allay allay ? allay.getJukebox() : null;
```

File: src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+	static {
+		registerDefault(ExprDuplicateCooldown.class, Timespan.class, "(duplicat(e|ing|ion)|clon(e|ing)) cool[ ]down [time]", "livingentities");
+	}
+
+	@Override
+	public @Nullable Timespan convert(LivingEntity entity) {
+		if (!(entity instanceof Allay allay))
+			return null;
+		return new Timespan(TimePeriod.TICK, allay.getDuplicationCooldown());
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java:43]:
```suggestion
	public @Nullable Timespan convert(LivingEntity entity) {
		return entity instanceof Allay allay
			? new Timespan(TimePeriod.TICK, allay.getDuplicationCooldown())
			: return null;
```

File: src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+				long value = Math2.fit(0, current - ticks, Long.MAX_VALUE);
+				allay.setDuplicationCooldown(value);
+			};
+			case RESET -> Allay::resetDuplicationCooldown;
+			default -> throw new IllegalStateException("Unexpected value: " + mode);
+		};
+		for (LivingEntity entity : getExpr().getArray(event)) {
+			if (!(entity instanceof Allay allay))
+				continue;
+			consumer.accept(allay);
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java:75]:
```suggestion
			if (entity instanceof Allay allay)
				consumer.accept(allay);
```

File: src/test/skript/tests/syntaxes/effects/EffAllayDancing.sk
Comment (cheeezburga) [inline src/test/skript/tests/syntaxes/effects/EffAllayDancing.sk:1]: Other entities too?

File: src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk [hunk 8]
```diff
@@ -0,0 +1,32 @@
+test "duplicate cooldown":
+	###
+		Combines:
+			ExprDuplicateCooldown
+			EffAllayDuplicate
+			EffAllayCanDuplicate
+			CondAllayCanDuplicate
+	###
```
Comment (cheeezburga) [inline src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk:8]:
```suggestion
# includes ExprDuplicateCooldown, EffAllayDuplicate, EffAllayCanDuplicate, CondAllayCanDuplicate
test "duplicate cooldown":
```

File: src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk
Comment (cheeezburga) [inline src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk:1]: Tests for other entities, infinity, NaN, going below 0, going above Long.MAX_VALUE

File: src/main/java/ch/njol/skript/effects/EffAllayDuplicate.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import ch.njol.util.Kleenean;
+import org.bukkit.entity.Allay;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Make Allay Duplicate")
+@Description({
+	"Make an allay duplicate itself.",
+	"This effect will always make an allay duplicate regardless if the duplicate attribute is disabled."
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffAllayDuplicate.java:20]:
```suggestion
	"This effect will always make an allay duplicate irregardless of whether the duplicate attribute is disabled."
```

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		SyntaxStringBuilder builder = new SyntaxStringBuilder(event, debug);
+		builder.append("make", entities);
+		if (start) {
+			builder.append("start");
+		} else {
+			builder.append("stop");
+		}
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffAllayDancing.java:84]:
```suggestion
		builder.append(start ? "start" : "stop");
```

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+			//noinspection unchecked
+			Expression<Location> location = (Expression<Location>) exprs[2];
+			this.location = Direction.combine(direction, location);
+		}
+		return true;
+	}
+
+	@Override
+	protected void execute(Event event) {
+		Location location = this.location == null ? null : this.location.getSingle(event);
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/effects/EffAllayDancing.java:60]: Just like the other one, `this.location.getSingle(event)` might still return null even if `this.location` is set. Another good place for a runtime error, but again, probably just wait for the highlighting PR to be merged first.

**Review by cheeezburga — CHANGES_REQUESTED** [review]
Some things here and there, but looking good. I'm not sure I love the class names being specific to allays, not a big deal though

File: src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk
Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk:1]: Can  you even use `infinity` and `NaN` with timespan?

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java (hunk 10 above)
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/effects/EffAllayDancing.java:84]: I was told not to use ternaries with the SSB.

File: src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk
Comment (cheeezburga) [reply src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk:1]: Oh my bad, I forgot acceptChange returned that. Just some big timespans then, which have more ticks than Long.MAX_VALUE

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java (hunk 10 above)
Comment (cheeezburga) [reply src/main/java/ch/njol/skript/effects/EffAllayDancing.java:84]: Thats fair I suppose, the if else here just looks meh to me

File: src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java (hunk 6 above)
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java:43]: I would rather not use a ternary here

File: src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk
Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk:1]: So this requested change is not needed.
Long.MAX_VALUE is 9.22 quintillion.
1) This would be equivalent to 14.6 trillion years
2) Would this not simply break the Timespan since it is also stored and retrieved as a long?

File: src/main/java/ch/njol/skript/effects/EffAllayCanDuplicate.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+	"allow all allays to duplicate",
+	"disallow all allays to duplicate"
+})
+@Since("INSERT VERSION")
+public class EffAllayCanDuplicate extends Effect {
+
+	static {
+		Skript.registerEffect(EffAllayCanDuplicate.class,
+			"allow %livingentities% to (duplicate|clone)",
+			"disallow %livingentities% to (duplicate|clone)",
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/effects/EffAllayCanDuplicate.java:32]:
```suggestion
```
bad grammar

File: src/main/java/ch/njol/skript/effects/EffAllayCanDuplicate.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+	protected void execute(Event event) {
+		for (LivingEntity entity : entities.getArray(event)) {
+			if (entity instanceof Allay allay)
+				allay.setCanDuplicate(duplicate);
+		}
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return (duplicate ? "allow " : "disallow ") + entities.toString(event, debug) + " to duplicate";
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/effects/EffAllayCanDuplicate.java:57]:
```suggestion
		return (duplicate ? "allow " : "disallow ") + entities.toString(event, debug) + " to duplicate";
```
should be adjusted to use the pattern with proper grammar

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		//noinspection unchecked
+		entities = (Expression<LivingEntity>) exprs[0];
+		start = matchedPattern == 0;
+		if (start && exprs[1] != null) {
+			//noinspection unchecked
+			Expression<Direction> direction = (Expression<Direction>) exprs[1];
+			//noinspection unchecked
+			Expression<Location> location = (Expression<Location>) exprs[2];
+			this.location = Direction.combine(direction, location);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/effects/EffAllayDancing.java:53]: no need for variables here

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+		return true;
+	}
+
+	@Override
+	protected void execute(Event event) {
+		Location location = null;
+		if (this.location != null) {
+			location = this.location.getSingle(event);
+			//if (location == null)
+				// Runtime warning;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/effects/EffAllayDancing.java:64]:
```suggestion
```

File: src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+				long value = Math2.fit(0, current - ticks, Long.MAX_VALUE);
+				allay.setDuplicationCooldown(value);
+			};
+			case RESET -> Allay::resetDuplicationCooldown;
+			default -> throw new IllegalStateException("Unexpected value: " + mode);
+		};
+		for (LivingEntity entity : getExpr().getArray(event)) {
+			if (entity instanceof Allay allay)
+				consumer.accept(allay);
+		}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprDuplicateCooldown.java:75]: prefer not to have a consumer here

File: src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+		set {_entity} to entity
+
+	set the duplication cooldown of {_entity} to 5 seconds
+	assert the duplication cooldown of {_entity} is 5 seconds with "Duplication cooldown was not set"
+	add 3 seconds to the duplication cooldown of {_entity}
+	assert the duplication cooldown of {_entity} is 8 seconds with "Duplication cooldown did not have 3 seconds added"
+	remove 5 seconds from the duplication cooldown of {_entity}
+	assert the duplication cooldown of {_entity} is 3 seconds with "Duplication cooldown did not have 5 seconds removed"
+	clear the duplication cooldown of {_entity}
+	assert the duplication cooldown of {_entity} is 0 seconds with "Duplication cooldown time was not cleared"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk:13]: Needs to test removing more seconds that the cooldown currently is, as well as setting to a negative value, and adding a negative amount of seconds that would make the cooldown negative.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+	set the duplication cooldown of {_entity} to 1 hour
+	remove 2 hours from the duplication cooldown of {_entity}
+	assert the duplication cooldown of {_entity} is 0 seconds with "Duplication cooldown should be 0 seconds after removing more than current"
+	add (2 seconds * -1) to the duplication cooldown of {_entity}
+	assert the duplication cooldown of {_entity} is 0 seconds with "Duplication cooldown should be 0 seconds after adding a negative"
+	set the duplication cooldown of {_entity} to (1 minute * -1)
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/expressions/ExprDuplicateCooldown.sk:19]: I forgot that this accepts timespans, so these can be removed. Timespans will not go below 0, 1 minute * -1 is 0 minutes, so these aren't actually testing anything.

**Review by sovdeeth — CHANGES_REQUESTED** [review]
looking good

File: src/main/java/ch/njol/skript/effects/EffAllayDancing.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+	"Make an allay start or stop dancing.",
+	"By providing a location when making an allay dance, checks to see if the block at the location is a jukebox and playing music.",
+	"By not providing a location, the allay will dance forever."
+})
+@Examples({
+	"if last spawned allay is not dancing:",
+		"\tmake last spawned allay start dancing"
+})
+@Since("INSERT VERSION")
+public class EffAllayDancing extends Effect {
```
[the remark this answers is not in the document]
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/effects/EffAllayDancing.java:31]: It's reasonable to add it, I've done it for other PRs, just didn't see that Piglins could before.

[the remark this answers is not in the document]
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/effects/EffAllayDancing.java:31]: Just a note, it does fail the test, not sure why. I have tested manually with `skriptTestDev` to ensure the piglin does dance and `is dancing` works, which it does.
But I will leave out the test.

File: src/main/java/ch/njol/skript/effects/EffDancing.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+import org.bukkit.entity.Piglin;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+@Name("Dance")
+@Description({
+	"Make an allay or piglin start or stop dancing.",
+	"Providing a location for allays, checks to see if the the block at the location is a jukebox and playing music. "
+		+ "Not providing a location, the allay will dance forever.",
+	"Providing a timespan for piglins, makes them dance for that period of time."
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/effects/EffDancing.java:25]:
```suggestion
	"Providing a location only applies to allays. They will check to see if the the block at the location is a jukebox and playing music. "
		+ "If it isn't, they will stop dancing. If no location is provided, the allay will dance indefinitely.",
	"Providing a timespan only applies for piglins. It determines the length of time they will dance for. " 
		+ "If no timespan is provided, they will dance indefinitely."
```

File: src/main/java/ch/njol/skript/effects/EffDancing.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		//noinspection unchecked
+		entities = (Expression<LivingEntity>) exprs[0];
+		start = matchedPattern == 0;
+		if (start && exprs[1] != null) {
+			//noinspection unchecked
+			location = Direction.combine((Expression<Direction>) exprs[1], (Expression<Location>) exprs[2]);
+		}
+		if (parseResult.hasTag("for")) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/effects/EffDancing.java:62]: can we use a more descriptive tag please

File: src/test/skript/tests/syntaxes/effects/EffDancing.sk [hunk 22]
```diff
@@ -0,0 +1,15 @@
+test "allay dancing":
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/effects/EffDancing.sk:1]: piglin tests?

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/test/skript/tests/syntaxes/effects/EffDancing.sk (hunk 22 above)
Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/effects/EffDancing.sk:1]: I left a comment on your initial request for piglins explaining why there were no tests

Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/effects/EffDancing.sk:1]: Well, it failing is a bit concerning. How were you testing?

Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/effects/EffDancing.sk:1]: Exact same test as the allay. Just spawn a piglin instead and change the fail messages. I've tested manually, the exact same way as done through the test and it works.

Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/effects/EffDancing.sk:1]: By manually, do you mean running that exact trigger? or running a similar trigger? or using effect commands?

Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/effects/EffDancing.sk:1]: The exact code in a test command didn't fail? If that does fail, the test shouldn't either. Perhaps it's a chunk loading thing like we saw a while back

Comment (TheAbsolutionism) [reply src/test/skript/tests/syntaxes/effects/EffDancing.sk:1]: My bad, the exact code doesn't work, the code but with added `wait`s does work.
It takes 2 ticks for the `Piglin#isDancing` to reflect the updated dancing state. Atleast for manually updating, not sure for natural/vanilla

https://github.com/user-attachments/assets/ed3ea740-be8b-4377-b6fd-c8172de7432d

**Review by sovdeeth — APPROVED** [review]
